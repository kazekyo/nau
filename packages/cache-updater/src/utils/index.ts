export * from './getNodesFromConnection';
export * from './graphqlAST';
export * from './nonNullable';
