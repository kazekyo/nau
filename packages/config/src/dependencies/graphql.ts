/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import * as _GraphQL from 'graphql';
const mod = require.main || module;

let GraphQLModule;
try {
  console.log('-----try----');
  GraphQLModule = mod.require('graphql');
} catch {
  console.log('------cache------');
  GraphQLModule = _GraphQL;
}

export const {
  BREAK,
  GraphQLError,
  parseType,
  visit,
  isNonNullType,
  valueFromAST,
  isTypeSubTypeOf,
  getNullableType,
  typeFromAST,
  GraphQLNonNull,
  GraphQLObjectType,
  visitWithTypeInfo,
  isInputType,
  TypeInfo,
  isObjectType,
} = GraphQLModule as typeof _GraphQL;
