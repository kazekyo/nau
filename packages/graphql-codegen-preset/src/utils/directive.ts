export const ARGUMENT_DEFINITIONS_DIRECTIVE_NAME = 'argumentDefinitions';
export const ARGUMENTS_DIRECTIVE_NAME = 'arguments';
export const REFETCHABLE_DIRECTIVE_NAME = 'refetchable';

export const customDirectives = {
  arguments: 'directive @arguments on FRAGMENT_SPREAD',
  argumentDefinitions: 'directive @argumentDefinitions on FRAGMENT_DEFINITION',
  refetchable: 'directive @refetchable(queryName: String) on FRAGMENT_DEFINITION',
};
