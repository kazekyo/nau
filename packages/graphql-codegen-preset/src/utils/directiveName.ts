export const ARGUMENT_DEFINITIONS_DIRECTIVE_NAME = 'argumentDefinitions';
export const ARGUMENTS_DIRECTIVE_NAME = 'arguments';
export const REFETCHABLE_DIRECTIVE_NAME = 'refetchable';
