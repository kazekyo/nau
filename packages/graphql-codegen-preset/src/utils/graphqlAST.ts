import { Types } from '@graphql-codegen/plugin-helpers';
import { DirectiveName } from '@nau/core';
import {
  DirectiveNode,
  DocumentNode,
  FieldNode,
  FragmentDefinitionNode,
  Kind,
  OperationDefinitionNode,
  SelectionNode,
  SelectionSetNode,
} from 'graphql';
import { nonNullable } from './nonNullable';

export const pageInfoField: FieldNode = {
  kind: Kind.FIELD,
  name: { kind: Kind.NAME, value: 'pageInfo' },
  selectionSet: {
    kind: Kind.SELECTION_SET,
    selections: [
      { kind: Kind.FIELD, name: { kind: Kind.NAME, value: 'hasNextPage' } },
      { kind: Kind.FIELD, name: { kind: Kind.NAME, value: 'endCursor' } },
      { kind: Kind.FIELD, name: { kind: Kind.NAME, value: 'hasPreviousPage' } },
      { kind: Kind.FIELD, name: { kind: Kind.NAME, value: 'startCursor' } },
    ],
  },
};

export const cursorField: FieldNode = { kind: Kind.FIELD, name: { kind: Kind.NAME, value: 'cursor' } };

export const idField: FieldNode = { kind: Kind.FIELD, name: { kind: Kind.NAME, value: 'id' } };

export const typenameField: FieldNode = { kind: Kind.FIELD, name: { kind: Kind.NAME, value: '__typename' } };

export const nodeField: FieldNode = {
  kind: Kind.FIELD,
  name: { kind: Kind.NAME, value: 'node' },
  selectionSet: {
    kind: Kind.SELECTION_SET,
    selections: [idField, typenameField],
  },
};

export const edgesField: FieldNode = {
  kind: Kind.FIELD,
  name: { kind: Kind.NAME, value: 'edges' },
  selectionSet: {
    kind: Kind.SELECTION_SET,
    selections: [nodeField, cursorField],
  },
};

export const connectionIdField: FieldNode = {
  kind: Kind.FIELD,
  name: { kind: Kind.NAME, value: '_connectionId' },
  directives: [{ kind: Kind.DIRECTIVE, name: { kind: Kind.NAME, value: 'client' } }],
};

export function getOperationDefinitions(doc: DocumentNode): OperationDefinitionNode[] {
  return doc.definitions.filter((definition) => definition.kind === 'OperationDefinition') as OperationDefinitionNode[];
}

export const getFragmentDefinitions = (documentNode: DocumentNode): FragmentDefinitionNode[] => {
  return documentNode.definitions.filter(
    (definition): definition is FragmentDefinitionNode => definition.kind === 'FragmentDefinition',
  );
};

export const getFragmentDefinitionsByDocumentFiles = (
  documentFiles: Types.DocumentFile[],
): FragmentDefinitionNode[] => {
  return documentFiles
    .map((file) => file.document?.definitions)
    .filter(nonNullable)
    .flat()
    .filter((definition): definition is FragmentDefinitionNode => definition.kind === Kind.FRAGMENT_DEFINITION);
};

export const getFragmentDefinitionByName = ({
  fragmentDefinitions,
  fragmentName,
}: {
  fragmentDefinitions: FragmentDefinitionNode[];
  fragmentName: string;
}): FragmentDefinitionNode | undefined => {
  return fragmentDefinitions.find((definition) => definition.name.value === fragmentName);
};

export const getDirectives = ({
  node,
  directiveNames,
}: {
  node: FieldNode | SelectionNode | null;
  directiveNames: DirectiveName[];
}): DirectiveNode[] => {
  const directives = node?.directives?.filter((directive) =>
    directiveNames.find((name) => name === directive.name.value),
  );
  if (!directives) return [];
  return directives;
};

export const isSameNameFieldNode = ({ selection, name }: { selection: SelectionNode; name: string }): boolean => {
  return selection.kind === Kind.FIELD && selection.name.value === name;
};

// Add the field, but do nothing if the field already exists
export const addFieldWithoutDuplication = ({
  fieldNode,
  additionalFields,
}: {
  fieldNode: FieldNode;
  additionalFields: FieldNode[];
}): FieldNode => {
  if (!fieldNode.selectionSet) return fieldNode;
  const selectionSet = addFieldToSelectionSetNodeWithoutDuplication({
    selectionSetNode: fieldNode.selectionSet,
    additionalFields,
  });

  return {
    ...fieldNode,
    selectionSet,
  };
};

// Add the field, but do nothing if the field already exists
export const addFieldToSelectionSetNodeWithoutDuplication = ({
  selectionSetNode,
  additionalFields,
}: {
  selectionSetNode: SelectionSetNode;
  additionalFields: FieldNode[];
}): SelectionSetNode => {
  const selections = selectionSetNode.selections;

  const fieldNodes = additionalFields.filter(
    (fieldNode) => !selections.find((selection) => isSameNameFieldNode({ selection, name: fieldNode.name.value })),
  );
  if (fieldNodes.length === 0) return selectionSetNode;

  return {
    ...selectionSetNode,
    selections: [...selections, ...fieldNodes],
  };
};
