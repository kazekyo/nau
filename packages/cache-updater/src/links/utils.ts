import { ApolloLink, FetchResult, Operation } from '@apollo/client';
import { Observable } from '@apollo/client/utilities';
import { VariableDefinitionNode } from 'graphql/language';
import { isSubscriptionOperation } from '../utils';

export type ConnectionArgumentDataType = { node: VariableDefinitionNode; name: string };
export type ConnectionArgumentsDataType = {
  first?: ConnectionArgumentDataType;
  last?: ConnectionArgumentDataType;
  after?: ConnectionArgumentDataType;
  before?: ConnectionArgumentDataType;
};
export type ContextType = {
  nau?: {
    refetch?: { fragmentName: string };
    connection?: { variables?: { count: number; cursor: string }; argumentsData?: ConnectionArgumentsDataType };
  };
};

export const createApolloLink = (transform: (operation: Operation) => Operation): ApolloLink => {
  return new ApolloLink((operation, forward) => {
    if (!forward) return null;

    const newOperation = transform(operation);
    if (isSubscriptionOperation(newOperation.query)) {
      return new Observable<FetchResult>((observer) =>
        forward(newOperation).subscribe((response) => observer.next(response)),
      );
    }

    return forward(newOperation).map((response) => response);
  });
};
