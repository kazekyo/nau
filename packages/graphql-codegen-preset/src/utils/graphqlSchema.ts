import { GraphQLObjectType, GraphQLOutputType, GraphQLSchema, isListType, isNonNullType, isObjectType } from 'graphql';
import { getFieldDef } from 'graphql/execution/execute';
import { Maybe } from 'graphql/jsutils/Maybe';
import { edgesField, nodeField } from './graphqlAST';

export const getConnectionType = ({ type }: { type: Maybe<GraphQLOutputType> }): GraphQLObjectType | undefined => {
  let connectionType = type;
  if (!connectionType) return;
  if (isNonNullType(connectionType)) {
    connectionType = connectionType.ofType as GraphQLObjectType;
  }
  if (!isObjectType(connectionType)) return;
  if (!connectionType.name.endsWith('Connection')) return;
  return connectionType;
};

export const getEdgeType = ({
  connectionType,
  schema,
}: {
  connectionType: GraphQLObjectType;
  schema: GraphQLSchema;
}): GraphQLObjectType | undefined => {
  const edgesFieldDef = getFieldDef(schema, connectionType, edgesField);
  if (!edgesFieldDef) return;

  let edgesType = edgesFieldDef.type;
  if (isNonNullType(edgesType)) {
    edgesType = edgesType.ofType;
  }
  if (!isListType(edgesType)) return;

  let edgeType = edgesType.ofType as Maybe<unknown>;
  if (!edgeType) return;

  if (isNonNullType(edgeType)) {
    edgeType = edgeType.ofType;
  }
  if (!isObjectType(edgeType)) return;

  return edgeType;
};

export const getNodeType = ({
  edgeType,
  schema,
}: {
  edgeType: GraphQLObjectType;
  schema: GraphQLSchema;
}): GraphQLObjectType | undefined => {
  const nodeFieldDef = getFieldDef(schema, edgeType, nodeField);
  if (!nodeFieldDef) return;

  let nodeType = nodeFieldDef.type;
  if (!nodeType) return;

  if (isNonNullType(nodeType)) {
    nodeType = nodeType.ofType;
  }
  if (!isObjectType(nodeType)) return;

  return nodeType;
};
