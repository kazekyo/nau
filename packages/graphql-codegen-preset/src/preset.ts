import { Types } from '@graphql-codegen/plugin-helpers';
import { checkValidationErrors, validateGraphQlDocuments } from '@graphql-tools/utils';
import {
  RelayArgumentsOfCorrectType,
  RelayDefaultValueOfCorrectType,
  RelayKnownArgumentNames,
  RelayNoUnusedArguments,
} from '@relay-graphql-js/validation-rules';
import { ASTVisitor, buildASTSchema, GraphQLSchema, specifiedRules, ValidationContext, ValidationRule } from 'graphql';
import cloneDeep from 'lodash.clonedeep';
import * as paginationPlugin from './plugins/cache-updater-support';
import { PresetConfig } from './presetConfig';
import { addCustomClientDirective } from './schemaTransforms/addClientDirective';
import { addConnectionId } from './schemaTransforms/addConnectionId';
import { transform as addFieldsForAddingNode } from './transforms/addFieldsForAddingNode';
import { transform as addPaginationFields } from './transforms/addPaginationFields';
import { transform as fixVariableNotDefinedInRoot } from './transforms/fixVariableNotDefinedInRoot';
import { transform as generateRefetchQuery } from './transforms/generateRefetchQuery';
import { transform as passArgumentValueToFragment } from './transforms/passArgumentValueToFragment';
import { transform as removeCustomDirective } from './transforms/removeCustomDirective';
import { paginationDirectiveValidationRule } from './validationRules/paginationDirective';

const transformDocuments = ({
  documentFiles,
}: {
  documentFiles: Types.DocumentFile[];
}): { documentFiles: Types.DocumentFile[] } => {
  let result = { documentFiles };
  [
    passArgumentValueToFragment,
    generateRefetchQuery,
    fixVariableNotDefinedInRoot,
    addPaginationFields,
    addFieldsForAddingNode,
  ].forEach((transformFunc) => {
    result = transformFunc(result);
  });
  return result;
};

const transformSchema = (schema: GraphQLSchema, documentFiles: Types.DocumentFile[]): GraphQLSchema => {
  let result = schema;
  [addCustomClientDirective, addConnectionId].forEach((transform) => {
    result = transform(result, documentFiles).schema;
  });
  return result;
};

const validationRules = (): ValidationRule[] => {
  const ignored = [
    'NoUnusedFragmentsRule',
    'NoUnusedVariablesRule',
    'KnownArgumentNamesRule',
    'NoUndefinedVariablesRule',
  ];
  const v4ignored = ignored.map((rule) => rule.replace(/Rule$/, ''));

  const rules = specifiedRules.filter(
    (f: (context: ValidationContext) => ASTVisitor) => !ignored.includes(f.name) && !v4ignored.includes(f.name),
  );
  return [
    ...rules,
    RelayArgumentsOfCorrectType,
    RelayDefaultValueOfCorrectType,
    RelayNoUnusedArguments,
    RelayKnownArgumentNames,
    paginationDirectiveValidationRule,
  ];
};

export const preset: Types.OutputPreset<PresetConfig> = {
  buildGeneratesSection: (options) => {
    const originalGraphQLSchema: GraphQLSchema = options.schemaAst
      ? options.schemaAst
      : buildASTSchema(options.schema, options.config);

    const schemaObject = transformSchema(originalGraphQLSchema, options.documents);

    return validateGraphQlDocuments(schemaObject, options.documents, validationRules()).then((errors) => {
      checkValidationErrors(errors);

      const transformedObject = transformDocuments({ documentFiles: options.documents });

      let pluginMap = options.pluginMap;
      let plugins = options.plugins;
      const { generateTypeScriptCode } = options.presetConfig;
      if (generateTypeScriptCode) {
        pluginMap = { [`nau-pagination-code`]: paginationPlugin, ...pluginMap };
        plugins = [
          // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
          { [`nau-pagination-code`]: { documentFiles: cloneDeep(transformedObject.documentFiles) } },
          ...plugins,
        ];
      }

      const { documentFiles } = removeCustomDirective({ documentFiles: transformedObject.documentFiles });

      const result = [
        {
          filename: options.baseOutputDir,
          plugins: plugins,
          pluginMap: pluginMap,
          config: options.config,
          schema: options.schema,
          schemaAst: schemaObject,
          documents: documentFiles,
        },
      ];
      return result;
    });
  },
};
