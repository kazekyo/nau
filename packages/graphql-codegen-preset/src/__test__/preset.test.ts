import { executeCodegen } from '@graphql-codegen/cli';
import { Source } from '@graphql-tools/utils';
import { parse } from 'graphql';
import path from 'path';
import { preset } from '../preset';
import { printDocuments, testSchemaDocumentNode } from '../utils/testing/utils';

describe('preset', () => {
  it('transforms documents', async () => {
    const documents: Source[] = [
      {
        document: parse(/* GraphQL */ `
          query TestQuery($cursor: String) {
            viewer {
              id
              ...Fragment_user @arguments(count: 5)
            }
          }
        `),
      },
      {
        document: parse(/* GraphQL */ `
          fragment Fragment_user on User @argumentDefinitions(count: { type: "Int", defaultValue: 1 }) {
            items(first: $count, after: $cursor) @pagination {
              edges {
                node {
                  id
                }
              }
            }
          }
        `),
      },
    ];
    const expectedDocument = parse(/* GraphQL */ `
      query TestQuery($cursor: String) {
        viewer {
          id
          ...Fragment_user
        }
      }
      fragment Fragment_user on User {
        items(first: 5, after: $cursor) {
          edges {
            node {
              id
              __typename
            }
            cursor
          }
          pageInfo {
            hasNextPage
            endCursor
            hasPreviousPage
            startCursor
          }
        }
      }
    `);

    const result = await preset.buildGeneratesSection({
      baseOutputDir: 'src/generated/graphql.ts',
      config: {},
      presetConfig: {},
      schema: testSchemaDocumentNode,
      documents: documents,
      plugins: [],
      pluginMap: {},
    });

    expect(printDocuments(result[0].documents)).toBe(printDocuments([{ document: expectedDocument }]));
  });

  describe('generateTypeScriptCode', () => {
    it('outputs nothing code to the content when the flag is not set to true', async () => {
      const input = {
        schema: path.join(__dirname, '../utils/testing/example.graphql'),
        documents: path.join(__dirname, './fixtures/exampleFile.ts'),
        generates: {
          ['./generated.ts']: {
            preset,
            plugins: [],
          },
        },
      };

      const result = await executeCodegen(input);
      expect(result[0].content).toBe('');
    });

    it('outputs TypeScript code when the flag is true', async () => {
      const input = {
        schema: path.join(__dirname, '../utils/testing/example.graphql'),
        documents: path.join(__dirname, './fixtures/exampleFile.ts'),
        generates: {
          ['./generated.ts']: {
            preset,
            plugins: [],
            presetConfig: {
              generateTypeScriptCode: true,
            },
          },
        },
      };

      const result = await executeCodegen(input);
      expect(result[0].content).toMatchSnapshot();
    });
  });
});
