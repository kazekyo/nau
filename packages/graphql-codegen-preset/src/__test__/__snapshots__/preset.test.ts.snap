// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preset generateTypeScriptCode outputs TypeScript code when the flag is true 1`] = `
"import { TypePolicy } from '@apollo/client';
import { withCacheUpdaterInternal } from '@nau/cache-updater';

export const paginationMetaList = [{ node: { typename: 'Item' }, parents: [{ typename: 'User', connection: { fieldName: 'items' }, edge: { typename: 'ItemEdge' } }] }]

export type CacheUpdaterTypePolicies = {
  User: TypePolicy;
  [__typename: string]: TypePolicy;
};

export const withCacheUpdater = (typePolicies: CacheUpdaterTypePolicies) =>
  withCacheUpdaterInternal({
    paginationMetaList,
    deleteRecordMetaList: [],
    typePolicies,
  });"
`;
